chapter4.adoc

```diff
@@ -221,9 +224,81 @@ follows:
 ], config:{lanes: 1, hspace:1024}}
 ....
 
+=== MTT access permissions lookup process
+
+MTT access-permissions for a physical address PA in the context of a
+supervisor domain is ascertained as follows:
+
+1. Let _a_ be `mttp.ppn` x _PAGESIZE_, and let _i_ = LEVELS, where for mode
+`Smmtt34[rw]`, _LEVELS_ = 2 and for `Smmtt[46 | 56][rw]` LEVELS = 3; _PAGESIZE_
+is 2^12; MTT _NON_LEAF_PTE_SIZE_ = 8 bytes. The `mttp` register must be active,
+i.e., the effective privilege mode must be not-M-mode.
+
+2. Let _mpte_ be the value of the `MTT` table entry at address _a_ + _pa.pn[i]_
+ x _NON_LEAF_PTE_SIZE_. If accessing _mpte_ violates a PMA or PMP check, raise
+an access-fault exception corresponding to the original access type.
+
+3. If any bits or encodings that are reserved for future standard use are
+set within _mpte_, stop and raise an access-fault exception corresponding to
+the original access type.
+
+4. Otherwise, the _mpte_ is valid. If (_i_=0) or (_i_=1 and _mpte.type_ is not
+`MTT_L1_DIR`), go to step 5. Otherwise, the _mpte_ is a pointer to the next
+level of the `MTT`. Let _i_ = _i_-1. If _i_ < 0, stop and raise an access-fault
+exception corresponding to the original access type. Otherwise, let
+_a_ = _mpte.ppn_ x _PAGESIZE_ and go to step 2. Note that when _mpte.type_ =
+`MTT_L1_DIR`, the _mpte.ppn_ field is the value of the _mpte.info_ field.
+
+5. A leaf _mpte_ has been found. If any bits or encodings within _mpte.type_
+and _mpte.info_ that are reserved for future standard use, per
+<<Smmtt-rw-l2-encoding>> and <<Smmtt-ad-l2-encoding>>, are set within _mpte_,
+stop and raise an access-fault exception corresponding to the access type.
+
+6. The _mpte_ is a valid leaf _mpte_. Fetch the access-permissions for the
+physical address per the steps described below:
+
+* if _i_=1, and the _mpte.type_ field specifies it holds the access-permission
+encoding for 1GB pages per <<Smmtt-rw-l2-encoding>> and <<Smmtt-ad-l2-encoding>>
+; go to step 7, else
+
+* if _i_=1, and the _mpte.type_ field specifies it holds the access-permission
+encoding for 2MB pages; the _mpte.info_ field contains the access-permission
+encoding per <<Smmtt-rw-l2-encoding>> and <<Smmtt-ad-l2-encoding>>; go to
+step 7, else
+
+* If _i_=0, the _mpte_ contains a 2-bit entry for mode `Smmtt[34 | 46 | 56]`
+that holds the access-permission encodings (or 4-bit entry for mode
+`Smmtt[34 | 46 | 56]rw`) for 4KB pages. The 2/4-bit access-permission encoding
+for the PA is held in the ppn referenced by _mpte.info_ and indexed via
+_pa.pn[i]_. The encodings are specified in <<Smmtt-rw>> and <<Smmtt-ad>> for the
+`Smmtt[34 | 46 | 56]` and `Smmtt[34 | 46 | 56]rw` modes respectively.
```

Comment:
This could be more clearly specified if the *rw was separate into its own section or similar to LEVELS there should be a parameter for 2 or 4 bit that is setup ahead of time and then used in the algorithm to extract the right 2 or 4 bits - this could be described in the form of a shift and a mask.

Reply:
Could the two table formats have been unified as only 2 bits are needed in either case? 
Could it be:
00 - not allowed
01 - read and execute allowed
10 - read and write allowed
11 - read, write, and execute allowed
Then 01 and 10 encoding can just be reserved for non-rw mode but the table layout would be identical.

Reply:
will be updated in PR revision - ptal